### PR TITLE
[WIP] Api to downsample images when you instantiate them

### PR DESCRIPTION
**Pull request by iskakaushik**
The plan is to have some sort of a `DownsamplingImageProvider` on the dart side that accepts `maxWidth` and `maxHeight` parameters. It would then instantiate the codec with the said constraints.

This is meant to address a few issues around memory footprint of the images.

https://github.com/flutter/flutter/issues/26194

### CODE REVIEW

File: lib/ui/painting/codec.cc [hunk 1]
```diff
@@ -281,6 +349,18 @@ void InstantiateImageCodec(Dart_NativeArguments args) {
   const float decodedCacheRatioCap =
       tonic::DartConverter<float>::FromDart(Dart_GetNativeArgument(args, 3));
 
+  int maxWidth =
+      tonic::DartConverter<int>::FromDart(Dart_GetNativeArgument(args, 4));
+  if (maxWidth < 0) {
```
Comment (jonahwilliams) [inline lib/ui/painting/codec.cc]: Passing -1 and then swapping to the max int feels bad

Comment (iskakaushik) [reply lib/ui/painting/codec.cc]: 100% agree, will explore ways to pass in an optional value.

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -1649,16 +1649,21 @@ class Codec extends NativeFieldWrapperClass2 {
 /// failed.
 Future<Codec> instantiateImageCodec(Uint8List list, {
   double decodedCacheRatioCap = double.infinity,
+  int maxWidth = -1,
```
Comment (jonahwilliams) [inline lib/ui/painting.dart]: is maxWidth/Height the right parameter here? This API should probably take a Size instead. 

lets think about the behaviors:

- I want to render a small image into a larger area
- I want to render a large image into a smaller area
- I want to resize a large image into a smaller area
- I want to resize a small image into a larger area (?)

I'm also not sure if this should be tied to the existing imageCodec, would it make more sense to expose a separate method for resizing?

Also, this API doesn't allow us to decode an image without loading it entirely into memory - which was a point brought up on the original problem

**Comment by jonahwilliams** [comment]
I would like to see some more examples of how this addresses the points in flutter/flutter#26194. It doesn't need to solve every problem, but we should be explicit about how this would be used and what problems it does solve.

**Comment by iskakaushik** [comment]
* This does not solve the problem of decoding without loading the image into memory. I don't think there is a way to do that using Skia. It does get you the ability to generate thumbnails by passing in `maxWidth` and `maxHeight` parameters, though we'll have to load the image into memory once. Memory footprint will be proportional to the size of the scaled image, which is a win in a few cases.
* High dimension image in re-render in ListView while scroll up/down (https://github.com/flutter/flutter/issues/25469). The way I envision this solving this issue is: `Image.downSample(Image.asset('asset/${image}'), maxWidth: screenWidth)`. The problem here is as mentioned in https://github.com/flutter/flutter/issues/28682. The images were evicted from the cache which lead to us having to decode the images frequently. Not that we can down sample the images are reduce the memory footprint, cache eviction rate would reduce. https://github.com/flutter/flutter/issues/2848 would also be addressed the same approach.
* This as it is currently defined does not provide a way to resize the image, the main use case this tries to address is the memory pressure from large images on small widgets, which has been flagged in a few other issues too.

**Pull request closed without merging**